src/transaction/log_reader.hpp

```diff
@@ -0,0 +1,95 @@
+#ifndef LOG_READER_HPP
+#define LOG_READER_HPP
+
+#include <type_traits>
+
+#include "log_lsa.hpp"
+#include "log_impl.h"
+
+/* encapsulates reading of the log
+ *
+ * NOTE: not thread safe
+ */
+class log_reader final
+{
+  public:
+    log_reader ();
+    log_reader (log_reader const & ) = delete;
+    log_reader (log_reader && ) = delete;
+
+    log_reader &operator = (log_reader const & ) = delete;
+    log_reader &operator = (log_reader && ) = delete;
+
+    inline const log_lsa &get_lsa() const
+    {
+      return m_lsa;
+    }
+
+    inline std::int64_t get_pageid() const
+    {
+      return m_lsa.pageid;
+    }
+
+    inline std::int16_t get_offset() const
+    {
+      return m_lsa.offset;
+    }
+
+    int set_lsa_and_fetch_page (const log_lsa &lsa);
+    const log_hdrpage &get_page_header() const;
+
+    template <typename T>
+    const typename std::remove_const< typename std::remove_reference<T>::type >::type *reinterpret_cptr () const;
+
+    void add (size_t size);
```

Comment:
@ribram `add` is ONLY used in function `log_rv_get_unzip_log_data` (~535) can we just refactor that function such that we don't need `add` at all?
I'm thinking:
- replace `add` on `then` branch with `add_align`
- remove `align` from the end
- I am not sure about the `else` branch since calling `copy_from_log` will, possibly, advance the log page

Reply:
It looks safe to replace with `add_align`. If `align` is removed from the end, looks like it must be called after `copy_from_log`.

Reply:
Still..
`add_align` internally can advance the page if the offset is passed the `LOGAREA_SIZE`.
On the `then` branch `area_ptr` will get a pointer to the current page. Is there any chance that calling `add_align` here will advance the page? Suppose a corner case where adding the `length` to the offset falls in the last `DOUBLE_ALIGNMENT` in that page (f.i.: last <8 bytes). If the answer is yes, then we cannot `add_align` because `area_ptr` will point to invalid data.

Reply:
Indeed, `add_align` would advance the page, if the offset passed `LOGAREA_SIZE`. However, is_withing_page already checks that `offset + size < LOGAREA_SIZE`, so calling `add_align` should be safe.

Reply:
[discussed separately]
Actually not, because `add_align` can advance the log page before the data that `area_ptr` points to gets to be used.
Resolved differently - with a flag - in order to remove the `log_reader.add` function.